### PR TITLE
First pairing week

### PR DESCRIPTION
**Pull request by jerminatorhits**

### CODE REVIEW

File: lib/game_session.rb [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    @display = Display.new(@board)
+    @translator = setup_translator
+    @current_player = @player1
+    # binding.pry
+    self.start_game
+
+  end
+
+  def start_game
+    p "Welcome to TicTacToe! Player 1 is 'X' and Player 2 is 'O'."
```
Comment (malinatran) [inline lib/game_session.rb:20]: Is there a reason why you're using `p` here and `print` elsewhere? As opposed to `puts`?

File: lib/game_session.rb [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      column_length.each do |col|
+        translator[key] = [(row - 1), (col - 1)]
+        key = key + 1
+      end
+    end
+    translator
+  end
+
+  def switch_turns
+    @current_player != player1 ? @current_player = player1 : @current_player = player2
```
Comment (malinatran) [inline lib/game_session.rb:63]: `@current_player = (@current_player = player1) ? player2 : player1`

File: lib/rules.rb [hunk 3]
```diff
@@ -0,0 +1,88 @@
+require 'pry'
+
+class Rules
+  attr_reader :board
+
+  def initialize(board)
+    @board = board
+  end
+
+  def check_winner
```
Comment (malinatran) [inline lib/rules.rb:10]: The word "check" as a prefix for a method name is misleading and doesn't actually tell us the intention behind the method's behavior. If you're returning a value (say the winning marker), it's better to rename this as "getWinningMarker." If you're determining whether or not there is a win, call this "isWinningGame" or "isWin." Methods that return a boolean value should have a name that is in the form of a question, and helpful for dynamically typed languages.

File: lib/rules.rb [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    horizontal_win || vertical_win || diagonal_win
+  end
+
+  def horizontal_win
```
Comment (malinatran) [inline lib/rules.rb:14]: Method names should start with a verb.

File: lib/rules.rb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      target_index = target_index + 1
+    end
+
+    if array_win(array)
+      winner = array[0]
+    end
+    winner
+  end
+
+  def antidiagonal_win
```
Comment (malinatran) [inline lib/rules.rb:59]: Looks just like the method above it – how can you refactor? Keep it DRY!

File: lib/rules.rb [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+      target_index = target_index - 1
+    end
+
+    if array_win(array)
+      winner = array[0]
+    end
+    winner
+  end
+
+  def array_win(array)
```
Comment (malinatran) [inline lib/rules.rb:74]: Make this a verb and don't introduce data types in your method names, especially in the event that the implementation details may change.

File: spec/game_session_spec.rb [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+     _4_|_5_|_6_\n
+        |   |
+     _7_|_8_|_9_\n")
+  end
+
+  it "should mark the board's grid" do
+    expect(@game_session.update_board_grid(1, @player1)).to eq([['X', nil, nil], [nil, nil, nil], [nil, nil, nil]])
+  end
+
+  it "should mark the board's grid" do
```
Comment (malinatran) [inline spec/game_session_spec.rb:33]: Why are there two tests that essentially test the same thing?

File: spec/rules_spec.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+                   ['X', 'O', 'O'],
+                   ['X', 'O', nil]]
+    expect(@current_game.tie).to eq(false)
+  end
+
+  it "should check for a tie without empty spaces" do
+    @board.grid = [['O', 'O', 'X'],
+                   ['X', 'O', 'O'],
+                   ['X', 'O', 'X']]
+    expect(@current_game.tie).to eq(false)
```
Comment (malinatran) [inline spec/rules_spec.rb:56]: This should result in `true`. Also, there isn't a real need for this test since it's virtually testing the same behavior as line 38.

File: lib/game_session.rb [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  #run Rules.check_win
+  def current_board_status
```
Comment (malinatran) [inline lib/game_session.rb:67]: This class has a good amount of printing to standard out – I would consider moving this and UI-related methods to `Display`.

File: lib/board.rb [hunk 10]
```diff
@@ -0,0 +1,15 @@
+class Board
+
+  attr_accessor :row, :column, :grid
+# in future think about 1-D array?
+  def initialize(row, column)
```
Comment (malinatran) [inline lib/board.rb:5]: Why does a board need to take in two args? Isn't a board always proportional (e.g 3x3, 4x4, 5x5, etc.)?

File: lib/display.rb [hunk 11]
```diff
@@ -0,0 +1,23 @@
+class Display
+
+  attr_reader :board
+  attr_accessor :display_array
+
+  def initialize(board)
```
Comment (malinatran) [inline lib/display.rb:6]: Why pass in the board when you're going to use a hardcoded array? Is this just a placeholder value?

**Review by malinatran — COMMENTED** [review]
Overall, it looks like you two made significant process, so good job on your pairing session. The structure and organization of the game are pretty sound!

I would think more critically about the tests that you're writing (including covering all test cases) and let that drive how you build the game, rather than try to anticipate the game from the top-down. There seems to be some concerns with separation of concerns (pun!). After all, isn't it the `Player`'s responsibility to get the move? And `Display`'s to handle what gets output to the user? Some of the method names need to better reflect their intent. I would also clean up the code in small, but important, ways including: removing new lines after `end` statements, removing comments, etc. Beyond the code, clear commit messages and READMEs are helpful.

File: lib/game_session.rb [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+
+  def initialize(player1, player2, board, rules, display)
+    @player1 = Player.new("X")
+    @player2 = Player.new("O")
+    @board = Board.new(3, 3)
+    @rules = Rules.new(@board)
+    @display = Display.new(@board)
+    @translator = setup_translator
+    @current_player = @player1
+    # binding.pry
```
Comment (portatlas) [inline lib/game_session.rb:14]: Just as Tony mentioned we want to get rid of any debugging code

File: README.md [hunk 13]
```diff
@@ -1 +0,0 @@
-# new-ruby-tic-tac-toe
```
Comment (portatlas) [inline README.md:1]: Can you add a README.md similar to your refactoring kata?

File: spec/game_session_spec.rb [hunk 14]
```diff
@@ -0,0 +1,70 @@
+require 'game_session'
+require 'pry'
+
+describe GameSession do
```
Comment (portatlas) [inline spec/game_session_spec.rb:4]: When running the test "Welcome to TicTacToe! Player 1 is 'X' and Player 2 is '0'."  is displayed to the console. The test should run without any noise. Can we capture the output (https://stackoverflow.com/questions/14987362/how-can-i-capture-stdout-to-a-string) before it is sent to standard out. This should also make testing the UI easier.

File: lib/game_session.rb [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+  end
+
+  def start_game
+    p "Welcome to TicTacToe! Player 1 is 'X' and Player 2 is 'O'."
+    @display.display_board
+    # provide_move(@current_player) #comment for testing purposes
+  end
+
+  def provide_move(player)
+    print "Player #{player.mark}'s turn: "
```
Comment (portatlas) [inline lib/game_session.rb:26]: Is there a way we can remove the dependency to Ruby's print? Perhaps in the future, we want to repurpose the game to play over a browser instead of the console.

File: lib/game_session.rb [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+
+  end
+
+  def start_game
+    p "Welcome to TicTacToe! Player 1 is 'X' and Player 2 is 'O'."
+    @display.display_board
+    # provide_move(@current_player) #comment for testing purposes
+  end
+
+  def provide_move(player)
```
Comment (portatlas) [inline lib/game_session.rb:25]: Should the GameSession know about making a move or should that be a method for a Player?

File: lib/rules.rb [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+  end
+
+  def check_winner
+    horizontal_win || vertical_win || diagonal_win
+  end
+
+  def horizontal_win
+    winner = nil
+    board.grid.each do |row|
+      if array_win(row)
```
Comment (portatlas) [inline lib/rules.rb:17]: this conditional seems quite similar to the ones on line 33 and 53, could we refactor this out into its own private method?

File: lib/display.rb (hunk 11 above)
Comment (kzapata19) [reply lib/display.rb:6]: That board was part of an earlier version of Display that didn't get used and should've been deleted. It's irrelevant to the Display now. DELETED.

File: lib/game_session.rb (hunk 1 above)
Comment (kzapata19) [reply lib/game_session.rb:20]: display_board should output a neatly formatted stringified board (including the new lines to separate the rows). Using either print or puts will achieve this. Ill use print to remain consistent with the methods mentioned below.

start_game and provide_move will be using print (instead of p or puts) because the desired output should not include an appended new line after the string messages.

File: lib/game_session.rb (hunk 2 above)
Comment (kzapata19) [reply lib/game_session.rb:63]: I see what you did there. UPDATED

File: lib/game_session.rb (hunk 9 above)
Comment (kzapata19) [reply lib/game_session.rb:67]: That makes sense. Thanks!

File: lib/rules.rb (hunk 3 above)
Comment (kzapata19) [reply lib/rules.rb:10]: Renamed 'check_winner' to 'get_winning_mark' and the rest of win/tie methods to include 'is' prefix.

File: lib/rules.rb (hunk 4 above)
Comment (kzapata19) [reply lib/rules.rb:14]: Renamed methods to include the 'is' prefix.

File: lib/rules.rb (hunk 3 above)
Comment (kzapata19) [reply lib/rules.rb:10]: Removed 'is' prefix from win/tie methods because they don't return booleans, they return the winning mark. Renamed to use 'get' prefix and 'mark' suffix (e.g. 'get_horizontal_winner_mark')

File: lib/rules.rb (hunk 4 above)
Comment (kzapata19) [reply lib/rules.rb:14]: Same as above

File: lib/rules.rb (hunk 6 above)
Comment (kzapata19) [reply lib/rules.rb:74]: Renamed to 'get_mark(moves)'

File: spec/game_session_spec.rb (hunk 7 above)
Comment (kzapata19) [reply spec/game_session_spec.rb:33]: To ensure the method is placing the expected mark in expected positions. Probably a more useful test would consider edge cases like if the player submits non 'X' or 'O' strings like nil for example

File: spec/rules_spec.rb (hunk 8 above)
Comment (kzapata19) [reply spec/rules_spec.rb:56]: Yes, you're right. This test does not test anything new. DELETED

Comment (kzapata19) [reply spec/rules_spec.rb:56]: Oh! I just remembered the reasoning behind the tests related to this 'tie' method. The three scenarios I wanted to test:
(1) when board is full AND no wins ("is_tie" should return true)
(2) when board is not yet completely full AND no wins yet ("is_tie" should return false)
(3) when board is full AND a player has a win ("is_tie" should return false - the board.grid referenced above) 

The assumption I made about what makes a 'tie' is that the board should be full (no spaces with nil values) and have no winner. Since an empty space can potentially still leave an opportunity for a player to win, I made this assumption.

File: lib/board.rb (hunk 10 above)
Comment (kzapata19) [reply lib/board.rb:5]: Good point! Changed Board to take in one argument for size.
